azkaban-common/src/main/java/azkaban/executor/dispatcher/CandidateComparator.java

```diff
@@ -0,0 +1,187 @@
+      if (item != null){
+        totalWeight += item.getWeight();
+      }
+    }
+
+    return totalWeight;
+  }
+
+  /** function to actually calculate the scores for the two objects that are being compared.
+   *  the comparison follows the following logic -
+   *  1. if both objects are equal return 0 score for both.
+   *  2. if one side is null, the other side gets all the score.
+   *  3. if both sides are non-null value, both values will be passed to all the registered FactorComparators
+   *     each factor comparator will generate a result based off it sole logic the weight of the comparator will be
+   *     added to the wining side, if equal, no value will be added to either side.
+   *  4. final result will be returned in a Pair container.
+   *
+   * */
+  public Pair<Integer,Integer> getReult(T object1, T object2){
+    logger.info(String.format("start comparing '%s' with '%s',  total weight = %s ",
+        object1 == null ? "(null)" : object1.toString(),
+        object2 == null ? "(null)" : object2.toString(),
+        this.getTotalWeight()));
+
+    // short cut if object equals.
+    if (object1 ==  object2){
+      logger.info("Result : 0 vs 0 (equal)");
+      return new Pair<Integer,Integer>(0,0);
+    }
+
+    // left side is null.
+    if (object1 == null){
+      logger.info(String.format("Result : 0 vs %s (left is null)",this.getTotalWeight()));
+      return new Pair<Integer, Integer>(0,this.getTotalWeight());
+    }
+
+    // right side is null.
+    if (object2 == null){
+      logger.info(String.format("Result : %s vs 0 (right is null)",this.getTotalWeight()));
+      return new Pair<Integer, Integer>(this.getTotalWeight(),0);
+    }
+
+    // both side is not null,put them thru the full loop
+    int result1 = 0 ;
+    int result2 = 0 ;
+    Collection<FactorComparator<T>> comparatorList = this.factorComparatorList.values();
+    Iterator<FactorComparator<T>> mapItr = comparatorList.iterator();
+    while (mapItr.hasNext()){
+      FactorComparator<T> comparator = (FactorComparator<T>) mapItr.next();
+      int result = comparator.compare(object1, object2);
+      result1  = result1 + (result > 0 ? comparator.getWeight() : 0);
+      result2  = result2 + (result < 0 ? comparator.getWeight() : 0);
+      logger.info(String.format("[Factor: %s] compare result : %s (current score %s vs %s)",
+          comparator.getFactorName(), result, result1, result2));
+    }
+
+    // in case of same score, user differentiator to stabilize the result.
+    if (result1 == result2){
+      boolean result = this.differentiate(object1, object2);
+      logger.info("[Differentiator] differentiator chose " + (result?  object1.toString(): object2.toString()));
+      if (result) result1++; else result2++;
+    }
+
+    logger.info(String.format("Result : %s vs %s ",result1,result2));
+    return new Pair<Integer,Integer>(result1,result2);
+  }
+
+  @Override
+  public int compare(T o1, T o2) {
```

Comment:
In case you missed, please take a look at the dispatcher class and pay attention to the filter logic.  the Flow info will be used there to filter out the "invalid" executors, let me know if that addresses (or partially  addresses :) ) your concern. 

As per our design talk the Flows should be generic and should be treated the same and therefore not counted in the comparison  between the executors (the filter was also not talked during the design and I added it for good).  And per the current design I think it makes sense to keep the flow info apart when doing the comparison between two executors,  at the tome two executors enter the comparison logic, both of them have already passed the filtering logic, which means both are ideally suitable for executing the flow as per the definition,  so the comparator will try to finger out , based on a rule set defined by the system, when comparing executor A to B, which one is better ? and this is how the logic works.      For your concern, I think what you want (let me know if I get it wrong)  is  "for this specific Flow tell me  whether executor A or executor B suits better ?" I think this is different feature ask and was not what we planed, plus I think if we go that route it gives more chance for the user who defined the flow the flexibility to "pick" executors, which may not be what we want .